\section{Background}
Repeat sequences are ubiquitous in complex genomes such as those of
mice \cite{msc02:ini} and men \cite{int01:ini}. In addition to repeats
within one sequence, repeats between two sequences can be used for
constructing fast alignment algorithms \cite{del99:ali}. As a result,
there is a lot of interest in algorithms for detecting repeats in
strings \cite{gus97:alg}. 


\section{Description of Program}
The program uses a suffix tree for locating either repeats of a
predefined length or for identifying the longest repeat. Sets of
repeats are marked by the same background color. The text window can
be freely edited. In addition, three chunks of 
interesting prose are provided:
\begin{enumerate}
  \I The sequence of the first genome to be sequenced, that of the
  bacteriophage $\Phi$-X174, which encodes 10
  proteins in 5386 nucleotides \cite{air78:nuc}.
  \I The alcohol dehydrogenase (\textit{Adh}) region of
  \textit{Drosophila melanogaster}. In 1983 this was the first gene to
  be subjected to comparative sequencing \cite{kre83:nuc}. 
  \I The text of the paper by Watson \& Crick announcing the
  double helical structure of DNA \cite{wat53:mol}.
\end{enumerate}

\section{Tutorial}
\begin{itemize}
  \I Look for the longest repeat in the double helix paper.
  \I Look for the next longest repeat in that paper. Is the repeat
  length correlated with the phrase's importance?
\end{itemize}






